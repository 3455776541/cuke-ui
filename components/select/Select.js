import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import cls from "classnames";
import Input from "../input";
import { DownIcon } from "../icon";

export default class Select extends PureComponent {
  state = {
    selectedValue: this.props.defaultValue || this.props.value || "",
    visible: false
  };
  static defaultProps = {
    prefixCls: "cuke-select",
    onPanelVisibleChange: () => {},
    onChange: () => {}
  };
  static propTypes = {
    prefixCls: PropTypes.string.isRequired,
    onPanelVisibleChange: PropTypes.func,
    onChange: PropTypes.func,
    overlay: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.string,
      PropTypes.object
    ])
  };
  constructor(props) {
    super(props);
  }

  onOpenOptionPanel = () => {
    this.setState({ visible: true });
    this.props.onPanelVisibleChange(true);
  };

  onCloseOptionPanel = () => {
    setTimeout(() => {
      this.setState({ visible: false });
      this.props.onPanelVisibleChange(false);
    }, 120);
  };

  onChange = value => {
    this.setState({ selectedValue: value });
    this.props.onChange(value);
  };
  render() {
    const { visible } = this.state;
    const {
      prefixCls,
      className,
      disabled,
      placeholder,
      children,
      onPanelVisibleChange, //eslint-disable-line
      ...attr
    } = this.props;

    const { selectedValue } = this.state;

    return (
      <div className={cls(`${prefixCls}`, className)} {...attr}>
        <div
          className={cls(`${prefixCls}-inner`, {
            [`${prefixCls}-active`]: visible
          })}
        >
          <Input
            disabled={disabled}
            readonly
            placeholder={placeholder}
            className={cls(`${prefixCls}-input`)}
            value={selectedValue}
            onFocus={this.onOpenOptionPanel}
            onBlur={this.onCloseOptionPanel}
          />
          <DownIcon className={`${prefixCls}-arrow`} />
        </div>
        <div
          className={cls(`${prefixCls}-content`, {
            [`${prefixCls}-open`]: visible,
            [`${prefixCls}-close`]: !visible
          })}
        >
          {React.Children.map(children, (element, index) => {
            return React.cloneElement(element, {
              key: index,
              selectedValue,
              onChange: this.onChange
            });
          })}
        </div>
      </div>
    );
  }
}
