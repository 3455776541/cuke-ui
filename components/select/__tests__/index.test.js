import React from "react";
import assert from "power-assert";
import { render, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Select from "../index";

describe("<Select/>", () => {
  it("should render a <Select/> components", () => {
    const wrapper = render(
      <Select>
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find cuke-select classnames", () => {
    const wrapper = shallow(
      <Select>
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    assert(wrapper.find(".cuke-select").length >= 1);
  });

  it("should cannot click when disabled", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Select disabled onChange={onChange}>
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.find(".cuke-select").simulate("click");
    expect(onChange).not.toHaveBeenCalled();
  });

  it("should can trigger click when disabled", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Select onChange={onChange}>
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.find(".cuke-select-input").simulate("click");
    setTimeout(() => {
      expect(onChange).toHaveBeenCalled();
    });
  });

  it("should render init default value", () => {
    const wrapper = shallow(
      <Select defaultValue="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.find(".cuke-select").simulate("click");
    assert(wrapper.state().selectedValue === "黄瓜");
  });

  it("should render init value", () => {
    const wrapper = shallow(
      <Select value="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.find(".cuke-select").simulate("click");
    assert(wrapper.state().selectedValue === "黄瓜");
  });

  it("should trigger onChange when clicked", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Select.Option value="茄子" onChange={onChange}>
        茄子
      </Select.Option>
    );
    wrapper.find(".cuke-select-option").simulate("click");
    expect(onChange).toHaveBeenCalled();
  });

  it("should window cannot trigger click event when un mount", () => {
    const onClick = jest.fn();
    const wrapper = shallow(
      <Select value="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.unmount();
    window.onclick = () => onClick;
    expect(onClick).not.toHaveBeenCalled();
  });

  it("should panel cannot hidden click event when window click", () => {
    const wrapper = shallow(
      <Select value="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );

    window.onclick();
    assert(wrapper.state().visible === null);
    wrapper.find(".cuke-select-input").simulate("click");
    assert(wrapper.state().visible === true);
    window.onclick();
    assert(wrapper.state().visible === true);
  });

  it("should trigger panel visible change", () => {
    const onPanelVisibleChange = jest.fn();
    const wrapper = shallow(
      <Select value="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.find(".cuke-select-input").simulate("click");
    setTimeout(() => {
      expect(onPanelVisibleChange).toHaveBeenCalled();
    });
  });

  it("should cannot trigger panel visible change when disabled groups", () => {
    const onPanelVisibleChange = jest.fn();
    const wrapper = shallow(
      <Select value="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.find(".cuke-select-input").simulate("click");
    expect(onPanelVisibleChange).not.toHaveBeenCalled();
  });
});
