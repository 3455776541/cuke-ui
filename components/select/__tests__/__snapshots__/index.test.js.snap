// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select/> should render a <Select/> components 1`] = `
<div
  class="cuke-select"
>
  <div
    class="cuke-select-inner"
  >
    <input
      class="cuke-input cuke-select-input"
      placeholder=""
      readonly=""
      type="text"
      value=""
    />
    <svg
      class="cuke-select-arrow"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
  <div
    class="cuke-select-content cuke-select-close"
  >
    <div
      class="cuke-select-option"
    >
      黄瓜
    </div>
    <div
      class="cuke-select-option"
    >
      茄子
    </div>
    <div
      class="cuke-select-option"
    >
      番茄
    </div>
  </div>
</div>
`;
